File: ts/Scanner.ts
import Lox from "./lox.js";
import { Token, TokenType } from "./Token.js";

export default class Scanner {
  private static readonly keywords: { [key: string]: TokenType } = {};

  static {
    this.keywords["and"] = TokenType.AND;
    this.keywords["class"] = TokenType.CLASS;
    this.keywords["else"] = TokenType.ELSE;
    this.keywords["false"] = TokenType.FALSE;
    this.keywords["for"] = TokenType.FOR;
    this.keywords["fun"] = TokenType.FUN;
    this.keywords["if"] = TokenType.IF;
    this.keywords["nil"] = TokenType.NIL;
    this.keywords["or"] = TokenType.OR;
    this.keywords["print"] = TokenType.PRINT;
    this.keywords["return"] = TokenType.RETURN;
    this.keywords["super"] = TokenType.SUPER;
    this.keywords["this"] = TokenType.THIS;
    this.keywords["true"] = TokenType.TRUE;
    this.keywords["var"] = TokenType.VAR;
    this.keywords["while"] = TokenType.WHILE;
  }

  private readonly source: string;
  private readonly tokens: Token[] = [];
  private start = 0;
  private current = 0;
  private line = 1;

  constructor(source: string) {
    this.source = source;
  }

  scanTokens(): Token[] {
    while (!this.isAtEnd()) {
      this.start = this.current;
      this.scanToken();
    }

    this.tokens.push(new Token(TokenType.EOF, "", null, this.line));
    return this.tokens;
  }

  private scanToken(): void {
    const c = this.advance();

    switch (c) {
      case "(":
        return this.addToken(TokenType.LEFT_PAREN);
      case ")":
        return this.addToken(TokenType.RIGHT_PAREN);
      case "{":
        return this.addToken(TokenType.LEFT_BRACE);
      case "}":
        return this.addToken(TokenType.RIGHT_BRACE);
      case ",":
        return this.addToken(TokenType.COMMA);
      case ".":
        return this.addToken(TokenType.DOT);
      case "-":
        return this.addToken(TokenType.MINUS);
      case "+":
        return this.addToken(TokenType.PLUS);
      case ";":
        return this.addToken(TokenType.SEMICOLON);
      case "*":
        return this.addToken(TokenType.STAR);
      case "!":
        return this.addToken(this.match("=") ? TokenType.BANG_EQUAL : TokenType.BANG);
      case "=":
        return this.addToken(this.match("=") ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
      case "<":
        return this.addToken(this.match("=") ? TokenType.LESS_EQUAL : TokenType.LESS);
      case ">":
        return this.addToken(this.match("=") ? TokenType.GREATER_EQUAL : TokenType.GREATER);
      case "/":
        if (this.match("/")) {
          // A comment goes until the end of the line
          while (this.peek() != "\n" && !this.isAtEnd()) this.advance();
        } else {
          this.addToken(TokenType.SLASH);
        }

        break;

      case " ":
      case "\r":
      case "\t":
        // Ignore whitespace
        break;

      case "\n":
        this.line++;
        break;

      case '"':
        return this.string();

      default:
        if (this.isDigit(c)) return this.number();
        if (this.isAlpha(c)) return this.identifier();
        return Lox.error(this.line, "Unexpected character.");
    }
  }

  private identifier(): void {
    while (this.isAlphaNumeric(this.peek())) this.advance();

    const text = this.source.substring(this.start, this.current);

    this.addToken(Scanner.keywords[text] || TokenType.IDENTIFIER);
  }

  private number(): void {
    while (this.isDigit(this.peek())) this.advance();

    // Look for a fractional part
    if (this.peek() == "." && this.isDigit(this.peekNext())) {
      // Consume the `.`
      this.advance();

      while (this.isDigit(this.peek())) this.advance();
    }

    this.addToken(TokenType.NUMBER, parseFloat(this.source.substring(this.start, this.current)));
  }

  private string(): void {
    while (this.peek() != '"' && !this.isAtEnd()) {
      if (this.peek() == "\n") this.line++;
      this.advance();
    }

    if (this.isAtEnd()) return Lox.error(this.line, "Unterminated string.");

    // The closing `"`
    this.advance();

    // Trim the surrounding quotes
    const value = this.source.substring(this.start + 1, this.current - 1);
    this.addToken(TokenType.STRING, value);
  }

  private match(expected: string): boolean {
    if (this.isAtEnd()) return false;
    if (this.source.charAt(this.current) != expected) return false;

    this.current++;
    return true;
  }

  private peek(): string {
    if (this.isAtEnd()) return "\0";
    return this.source.charAt(this.current);
  }

  private peekNext(): string {
    if (this.current + 1 >= this.source.length) return "\0";
    return this.source.charAt(this.current + 1);
  }

  private isAlpha(c: string): boolean {
    return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z") || c == "_";
  }

  private isAlphaNumeric(c: string): boolean {
    return this.isAlpha(c) || this.isDigit(c);
  }

  private isDigit(c: string): boolean {
    return c >= "0" && c <= "9";
  }

  private isAtEnd(): boolean {
    return this.current >= this.source.length;
  }

  private advance(): string {
    return this.source.charAt(this.current++);
  }

  private addToken(tokenType: TokenType, literal: any = null) {
    const text = this.source.substring(this.start, this.current);
    this.tokens.push(new Token(tokenType, text, literal, this.line));
  }
}

File: zig/src/Scanner.zig
const std = @import("std");

start: [*]const u8,
current: [*]const u8,
line: usize = 1,

pub const TokenType = enum {
    // Single-character tokens
    left_paren,
    right_paren,
    left_brace,
    right_brace,
    comma,
    dot,
    minus,
    plus,
    semicolon,
    slash,
    star,

    // One or two character tokens
    bang,
    bang_equal,
    equal,
    equal_equal,
    greater,
    greater_equal,
    less,
    less_equal,

    // Literals
    identifier,
    string,
    number,

    // Keywords
    @"and",
    class,
    @"else",
    false,
    @"for",
    fun,
    @"if",
    nil,
    @"or",
    print,
    @"return",
    super,
    this,
    true,
    @"var",
    @"while",

    @"error",
    eof,
};

pub const Token = struct {
    type: TokenType,
    lexeme: []const u8,
    line: usize,
};

pub fn init(source: []const u8) @This() {
    return .{ .start = source.ptr, .current = source.ptr };
}

pub fn scanToken(self: *@This()) Token {
    self.skipWhitespace();
    self.start = self.current;

    if (self.isAtEnd()) return self.makeToken(.eof);

    const byte = self.advance();

    if (isAlpha(byte)) return self.identifier();
    if (isDigit(byte)) return self.number();

    return switch (byte) {
        '(' => self.makeToken(.left_paren),
        ')' => self.makeToken(.right_paren),
        '{' => self.makeToken(.left_brace),
        '}' => self.makeToken(.right_brace),
        ';' => self.makeToken(.semicolon),
        ',' => self.makeToken(.comma),
        '.' => self.makeToken(.dot),
        '-' => self.makeToken(.minus),
        '+' => self.makeToken(.plus),
        '/' => self.makeToken(.slash),
        '*' => self.makeToken(.star),
        '!' => self.makeToken(if (self.match('=')) .bang_equal else .bang),
        '=' => self.makeToken(if (self.match('=')) .equal_equal else .equal),
        '<' => self.makeToken(if (self.match('=')) .less_equal else .less),
        '>' => self.makeToken(if (self.match('=')) .greater_equal else .greater),
        '"' => self.string(),
        else => self.errorToken("Unexpected character."),
    };
}

fn isAtEnd(self: @This()) bool {
    return self.current[0] == 0;
}

fn advance(self: *@This()) u8 {
    const byte = self.current[0];
    self.current += 1;
    return byte;
}

fn peek(self: @This()) u8 {
    return self.current[0];
}

fn peekNext(self: @This()) u8 {
    if (self.isAtEnd()) return 0;
    return self.current[1];
}

fn match(self: *@This(), expected: u8) bool {
    if (self.isAtEnd()) return false;
    if (self.current[0] != expected) return false;
    self.current += 1;
    return true;
}

fn makeToken(self: @This(), tokenType: TokenType) Token {
    const length = @ptrToInt(self.current) - @ptrToInt(self.start);

    return .{ .type = tokenType, .line = self.line, .lexeme = self.start[0..length] };
}

fn errorToken(self: @This(), message: []const u8) Token {
    return .{ .type = .@"error", .line = self.line, .lexeme = message };
}

fn skipWhitespace(self: *@This()) void {
    while (true) {
        switch (self.peek()) {
            ' ', '\r', '\t' => _ = self.advance(),
            '\n' => {
                self.line += 1;
                _ = self.advance();
            },
            '/' => {
                if (self.peekNext() == '/') {
                    while (self.peek() != '\n' and !self.isAtEnd()) _ = self.advance();
                } else {
                    return;
                }
            },
            else => return,
        }
    }
}

fn checkKeyword(self: @This(), start: usize, rest: []const u8, tokenType: TokenType) TokenType {
    const length = @ptrToInt(self.current) - @ptrToInt(self.start);

    if (length == start + rest.len and
        std.mem.eql(u8, self.start[start..rest.len], rest)) return tokenType;

    return .identifier;
}

fn identifierType(self: @This()) TokenType {
    return switch (self.start[0]) {
        'a' => self.checkKeyword(1, "nd", .@"and"),
        'c' => self.checkKeyword(1, "lass", .class),
        'e' => self.checkKeyword(1, "else", .@"else"),
        'f' => if (@ptrToInt(self.current) - @ptrToInt(self.start) > 1) switch (self.start[1]) {
            'a' => self.checkKeyword(2, "lse", .false),
            'o' => self.checkKeyword(2, "r", .@"for"),
            'u' => self.checkKeyword(2, "n", .fun),
            else => .identifier,
        } else .identifier,
        'i' => self.checkKeyword(1, "if", .@"if"),
        'n' => self.checkKeyword(1, "il", .nil),
        'o' => self.checkKeyword(1, "r", .@"or"),
        'p' => self.checkKeyword(1, "rint", .print),
        'r' => self.checkKeyword(1, "eturn", .@"return"),
        's' => self.checkKeyword(1, "uper", .super),
        't' => if (@ptrToInt(self.current) - @ptrToInt(self.start) > 1) switch (self.start[1]) {
            'h' => self.checkKeyword(2, "is", .this),
            'r' => self.checkKeyword(2, "ue", .true),
            else => .identifier,
        } else .identifier,
        'v' => self.checkKeyword(1, "ar", .@"var"),
        'w' => self.checkKeyword(1, "hile", .@"while"),
        else => .identifier,
    };
}

fn identifier(self: *@This()) Token {
    while (isAlpha(self.peek()) or isDigit(self.peek())) _ = self.advance();
    return self.makeToken(self.identifierType());
}

fn number(self: *@This()) Token {
    while (isDigit(self.peek())) _ = self.advance();

    // Look for a fractional part
    if (self.peek() == '.' and isDigit(self.peekNext())) {
        // Consume the "."
        _ = self.advance();

        while (isDigit(self.peek())) _ = self.advance();
    }

    return self.makeToken(.number);
}

fn string(self: *@This()) Token {
    while (self.peek() != '"' and !self.isAtEnd()) {
        if (self.peek() == '\n') self.line += 1;
        _ = self.advance();
    }

    if (!self.isAtEnd()) return self.errorToken("Unterminated string.");

    // The closing quote
    _ = self.advance();
    return self.makeToken(.string);
}

fn isAlpha(byte: u8) bool {
    return (byte >= 'a' and byte <= 'z') or
        (byte >= 'A' and byte <= 'Z') or
        byte == '_';
}

fn isDigit(byte: u8) bool {
    return byte >= '0' and byte <= '9';
}
